File: examples/language-binding-plugin/python/test_plugin.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Atomic tests for the Python language-binding plugin example."""

from __future__ import annotations

from collections.abc import AsyncIterator
from dataclasses import dataclass
from typing import Any

import pytest
import pytest_asyncio
from main import DocumentationPlugin, component

import nemo_relay
from nemo_relay import llm, plugin, runtime_registrations, scope, subscribers, tools
from nemo_relay.runtime_registrations import RuntimeRegistrationKind


@dataclass
class ActivatedExample:
    implementation: DocumentationPlugin
    report: dict[str, Any]


class RecordingContext:
    """Records one registration per context method without touching global state."""

    def __init__(self) -> None:
        self.registrations: list[str] = []

    def __getattr__(self, name: str):
        if not name.startswith("register_"):
            raise AttributeError(name)

        def register(_registration_name: str, *_args: Any) -> None:
            self.registrations.append(name)

        return register


@pytest_asyncio.fixture
async def active_plugin(tmp_path: Any, monkeypatch: pytest.MonkeyPatch) -> AsyncIterator[ActivatedExample]:
    """Activate a fresh component and remove every owned registration afterward."""

    implementation = DocumentationPlugin()
    monkeypatch.chdir(tmp_path)
    monkeypatch.setenv("XDG_CONFIG_HOME", str(tmp_path))
    plugin.register("documentation-plugin", implementation)
    try:
        report = await plugin.initialize(component("enforce"))
        yield ActivatedExample(implementation, report)
    finally:
        await plugin.clear_async()
        plugin.deregister("documentation-plugin")


def test_validation_accepts_supported_mode() -> None:
    assert DocumentationPlugin().validate({"requests": {"mode": "enforce"}}) == []


def test_default_registration_control_is_disabled_and_valid() -> None:
    configuration = component("enforce").components[0].config

    assert configuration["registration_control"]["enabled"] is False
    assert DocumentationPlugin().validate(configuration) == []


def test_validation_rejects_unsupported_mode() -> None:
    diagnostics = DocumentationPlugin().validate({"requests": {"mode": "invalid"}})

    assert diagnostics[0]["code"] == "documentation-plugin.unsupported_mode"


def test_validation_rejects_wrong_type() -> None:
    diagnostics = DocumentationPlugin().validate({"requests": {"priority": "high"}})

    assert diagnostics[0]["code"] == "documentation-plugin.invalid_config"


def test_registration_rejects_a_duplicate_kind_and_missing_deregistration_is_false() -> None:
    plugin.register("documentation-plugin", DocumentationPlugin())
    try:
        with pytest.raises(RuntimeError):
            plugin.register("documentation-plugin", DocumentationPlugin())
        assert plugin.deregister("missing-documentation-plugin") is False
    finally:
        plugin.deregister("documentation-plugin")


@pytest.mark.parametrize(
    ("config", "field", "code"),
    [
        ({"tag": ""}, "tag", "documentation-plugin.invalid_tag"),
        ({"requests": {"header_name": ""}}, "requests.header_name", "documentation-plugin.invalid_header"),
        ({"requests": {"header_value": ""}}, "requests.header_value", "documentation-plugin.invalid_header"),
        ({"registration_control": {"kinds": []}}, "registration_control.kinds", "documentation-plugin.invalid_config"),
        (
            {"registration_control": {"registration_name": ""}},
            "registration_control.registration_name",
            "documentation-plugin.invalid_config",
        ),
        (
            {"registration_control": {"reason": ""}},
            "registration_control.reason",
            "documentation-plugin.invalid_config",
        ),
    ],
)
def test_validation_rejects_empty_required_strings(config: dict[str, Any], field: str, code: str) -> None:
    diagnostics = DocumentationPlugin().validate(config)

    assert any(item["code"] == code and item["field"] == field for item in diagnostics)


def test_validation_warns_about_unknown_field() -> None:
    diagnostics = DocumentationPlugin().validate({"unexpected": True})

    assert diagnostics[0]["level"] == "warning"
    assert diagnostics[0]["field"] == "unexpected"


def test_disabled_invalid_component_is_still_validated() -> None:
    plugin.register("documentation-plugin", DocumentationPlugin())
    try:
        report = plugin.validate(component("invalid", enabled=False))
        assert report["diagnostics"][0]["code"] == "documentation-plugin.unsupported_mode"
    finally:
        plugin.deregister("documentation-plugin")


def test_registers_each_safe_plugin_surface() -> None:
    context = RecordingContext()

    DocumentationPlugin().register(component("enforce").components[0].config, context)  # type: ignore[arg-type]

    assert set(context.registrations) == {
        "register_subscriber",
        "register_mark_sanitize_guardrail",
        "register_scope_sanitize_start_guardrail",
        "register_scope_sanitize_end_guardrail",
        "register_tool_sanitize_request_guardrail",
        "register_tool_sanitize_response_guardrail",
        "register_tool_conditional_execution_guardrail",
        "register_tool_request_intercept",
        "register_tool_execution_intercept",
        "register_llm_sanitize_request_guardrail",
        "register_llm_sanitize_response_guardrail",
        "register_llm_conditional_execution_guardrail",
        "register_llm_request_intercept",
        "register_llm_execution_intercept",
        "register_llm_stream_execution_intercept",
    }

    configuration = component("enforce").components[0].config
    configuration["registration_control"]["enabled"] = True
    context = RecordingContext()
    DocumentationPlugin().register(configuration, context)  # type: ignore[arg-type]
    assert "register_conditional_middleware_guardrail" in context.registrations


async def test_registration_control_is_owned_by_activation(tmp_path: Any, monkeypatch: pytest.MonkeyPatch) -> None:
    target = "documentation-controlled-subscriber"
    observed: list[str] = []
    monkeypatch.chdir(tmp_path)
    monkeypatch.setenv("XDG_CONFIG_HOME", str(tmp_path))
    subscribers.register(target, lambda event: observed.append(event.name))
    configuration = component("enforce")
    configuration.components[0].config["registration_control"]["enabled"] = True
    plugin.register("documentation-plugin", DocumentationPlugin())
    try:
        before = runtime_registrations.list_runtime_registrations({RuntimeRegistrationKind.SUBSCRIBER})
        assert any(item.effective_name == target for item in before)
        await plugin.initialize(configuration)
        baseline = len(observed)
        scope.event("registration-control-active")
        await subscribers.flush_async()
        assert len(observed) == baseline
        await plugin.clear_async()
        scope.event("registration-control-cleared")
        await subscribers.flush_async()
        assert observed[-1] == "registration-control-cleared"
    finally:
        await plugin.clear_async()
        plugin.deregister("documentation-plugin")
        subscribers.deregister(target)


async def test_activation_reports_no_diagnostics(active_plugin: ActivatedExample) -> None:
    assert active_plugin.report["diagnostics"] == []


async def test_tool_request_is_rewritten(active_plugin: ActivatedExample) -> None:
    result = await tools.execute(
        "safe_tool",
        {"value": 1},
        lambda args: nemo_relay.ToolExecutionResult(args, {"source": "application"}),
    )

    assert result.result == {"value": 1, "plugin_tag": "documentation"}
    assert result.annotation == {"source": "application"}


async def test_tool_policy_blocks_configured_tool(active_plugin: ActivatedExample) -> None:
    with pytest.raises(RuntimeError, match="guardrail rejected"):
        await tools.execute("dangerous_tool", {"value": 1}, lambda _args: pytest.fail("provider must not run"))


async def test_llm_request_is_rewritten(active_plugin: ActivatedExample) -> None:
    request = nemo_relay.LLMRequest({}, {"model": "allowed-model"})

    result = await llm.execute("allowed-model", request, lambda rewritten: {"headers": rewritten.headers})

    assert result["headers"]["x-nemo-relay-plugin"] == "documentation"


async def test_llm_policy_blocks_configured_model(active_plugin: ActivatedExample) -> None:
    request = nemo_relay.LLMRequest({}, {"model": "restricted-model"})

    with pytest.raises(RuntimeError, match="guardrail rejected"):
        await llm.execute("restricted-model", request, lambda _request: pytest.fail("provider must not run"))


async def test_llm_stream_is_transformed(active_plugin: ActivatedExample) -> None:
    request = nemo_relay.LLMRequest({}, {"model": "allowed-model"})

    async def provider(_request: Any) -> AsyncIterator[dict[str, int]]:
        yield {"chunk": 1}
        yield {"chunk": 2}

    stream = await llm.stream_execute("allowed-model", request, provider, lambda _chunk: None, lambda: {"done": True})
    chunks = [chunk async for chunk in stream]

    assert chunks == [
        {"chunk": 1, "plugin_stream": True},
        {"chunk": 2, "plugin_stream": True},
    ]


async def test_subscriber_observes_managed_call(active_plugin: ActivatedExample) -> None:
    await tools.execute("safe_tool", {"value": 1}, nemo_relay.ToolExecutionResult)
    await subscribers.flush_async()

    assert active_plugin.implementation.events


async def test_runtime_events_do_not_depend_on_request_rewriting(
    tmp_path: Any, monkeypatch: pytest.MonkeyPatch
) -> None:
    monkeypatch.chdir(tmp_path)
    monkeypatch.setenv("XDG_CONFIG_HOME", str(tmp_path))
    implementation = DocumentationPlugin()
    configuration = component("enforce")
    configuration.components[0].config["requests"]["enabled"] = False
    plugin.register("documentation-plugin", implementation)
    try:
        await plugin.initialize(configuration)
        await tools.execute("safe_tool", {"value": 1}, nemo_relay.ToolExecutionResult)
        await subscribers.flush_async()

        assert "documentation-plugin.request" in implementation.events
    finally:
        await plugin.clear_async()
        plugin.deregister("documentation-plugin")


async def test_teardown_removes_plugin_kind(tmp_path: Any, monkeypatch: pytest.MonkeyPatch) -> None:
    monkeypatch.chdir(tmp_path)
    monkeypatch.setenv("XDG_CONFIG_HOME", str(tmp_path))
    plugin.register("documentation-plugin", DocumentationPlugin())
    try:
        await plugin.initialize(component("enforce"))
        await plugin.clear_async()
        assert plugin.deregister("documentation-plugin") is True
        assert "documentation-plugin" not in plugin.list_kinds()
    finally:
        await plugin.clear_async()
        plugin.deregister("documentation-plugin")

File: examples/language-binding-plugin/node/test-plugin.mjs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import assert from 'node:assert/strict';
import { test } from 'node:test';

import { DEFAULT_CONFIG, config, documentationPlugin, isolateExampleEnvironment, plugin, relay } from './main.mjs';

function registeredCallbacks() {
  const callbacks = new Map();
  const context = new Proxy(
    {},
    {
      get(_target, method) {
        return (...args) => callbacks.set(method, args.at(-1));
      },
    },
  );
  documentationPlugin.register(structuredClone(DEFAULT_CONFIG), context);
  return callbacks;
}

async function withActivePlugin(run, pluginConfig = config('enforce')) {
  const restoreEnvironment = isolateExampleEnvironment();
  documentationPlugin.events.length = 0;
  try {
    plugin.register('documentation-plugin', documentationPlugin);
    const preflight = plugin.validate(pluginConfig);
    assert.deepEqual(preflight.diagnostics, []);
    const report = await plugin.initialize(pluginConfig);
    return await run(report);
  } finally {
    // Scope-end sanitizers run in Relay's queued publication path. Drain that
    // work before removing the callbacks that the active component owns.
    await relay.flushSubscribers();
    plugin.clear();
    plugin.deregister('documentation-plugin');
    restoreEnvironment();
  }
}

test('validation accepts a supported mode', () => {
  assert.deepEqual(documentationPlugin.validate({ requests: { mode: 'enforce' } }), []);
});

test('default registration control is disabled and valid', () => {
  assert.equal(DEFAULT_CONFIG.registration_control.enabled, false);
  assert.deepEqual(documentationPlugin.validate(structuredClone(DEFAULT_CONFIG)), []);
});

test('validation rejects an unsupported mode', () => {
  const diagnostics = documentationPlugin.validate({ requests: { mode: 'invalid' } });

  assert.equal(diagnostics[0].code, 'documentation-plugin.unsupported_mode');
});

test('validation rejects a wrong type', () => {
  const diagnostics = documentationPlugin.validate({ requests: { priority: 'high' } });

  assert.equal(diagnostics[0].code, 'documentation-plugin.invalid_config');
});

test('validation reports a non-object configuration', () => {
  const diagnostics = documentationPlugin.validate(null);

  assert.equal(diagnostics[0].code, 'documentation-plugin.invalid_config');
  assert.equal(diagnostics[0].field, undefined);
});

for (const [config, field, code] of [
  [{ tag: '' }, 'tag', 'documentation-plugin.invalid_tag'],
  [{ requests: { header_name: '' } }, 'requests.header_name', 'documentation-plugin.invalid_header'],
  [{ requests: { header_value: '' } }, 'requests.header_value', 'documentation-plugin.invalid_header'],
  [{ registration_control: { kinds: [] } }, 'registration_control.kinds', 'documentation-plugin.invalid_config'],
  [
    { registration_control: { registration_name: '' } },
    'registration_control.registration_name',
    'documentation-plugin.invalid_config',
  ],
  [{ registration_control: { reason: '' } }, 'registration_control.reason', 'documentation-plugin.invalid_config'],
]) {
  test(`validation rejects an empty ${field}`, () => {
    const diagnostics = documentationPlugin.validate(config);

    assert.ok(diagnostics.some((diagnostic) => diagnostic.code === code && diagnostic.field === field));
  });
}

test('validation warns about an unknown field', () => {
  const diagnostics = documentationPlugin.validate({ unexpected: true });

  assert.equal(diagnostics[0].level, 'warning');
  assert.equal(diagnostics[0].field, 'unexpected');
});

test('disabled invalid configuration is still validated', () => {
  const restoreEnvironment = isolateExampleEnvironment();
  plugin.register('documentation-plugin', documentationPlugin);
  try {
    const report = plugin.validate(config('invalid', false));

    assert.equal(report.diagnostics[0].code, 'documentation-plugin.unsupported_mode');
  } finally {
    plugin.deregister('documentation-plugin');
    restoreEnvironment();
  }
});

test('registers every safe plugin surface', () => {
  const registrations = registeredCallbacks();

  assert.deepEqual([...registrations.keys()].sort(), [
    'registerLlmConditionalExecutionGuardrail',
    'registerLlmExecutionIntercept',
    'registerLlmRequestIntercept',
    'registerLlmSanitizeRequestGuardrail',
    'registerLlmSanitizeResponseGuardrail',
    'registerLlmStreamExecutionIntercept',
    'registerMarkSanitizeGuardrail',
    'registerScopeSanitizeEndGuardrail',
    'registerScopeSanitizeStartGuardrail',
    'registerSubscriber',
    'registerToolConditionalExecutionGuardrail',
    'registerToolExecutionIntercept',
    'registerToolRequestIntercept',
    'registerToolSanitizeRequestGuardrail',
    'registerToolSanitizeResponseGuardrail',
  ]);

  const enabled = structuredClone(DEFAULT_CONFIG);
  enabled.registration_control.enabled = true;
  const methods = new Set();
  documentationPlugin.register(enabled, new Proxy({}, { get: (_target, method) => () => methods.add(method) }));
  assert.ok(methods.has('registerConditionalMiddlewareGuardrail'));
});

test('registration control is owned by activation', async () => {
  const target = 'documentation-controlled-subscriber';
  const observed = [];
  const controlled = config('enforce');
  controlled.components[0].config.registration_control.enabled = true;
  relay.registerSubscriber(target, (event) => observed.push(event.name));
  try {
    const before = relay.listRuntimeRegistrations(['subscriber']);
    assert.ok(before.some((registration) => registration.effectiveName === target));
    await withActivePlugin(async () => {
      const baseline = observed.length;
      relay.event('registration-control-active', null, null);
      await relay.flushSubscribers();
      assert.equal(observed.length, baseline);
    }, controlled);
    relay.event('registration-control-cleared', null, null);
    await relay.flushSubscribers();
    assert.equal(observed.at(-1), 'registration-control-cleared');
  } finally {
    relay.deregisterSubscriber(target);
  }
});

test('activation reports no diagnostics', async () => {
  await withActivePlugin((report) => {
    assert.deepEqual(report.diagnostics, []);
  });
});

test('tool requests are rewritten', async () => {
  await withActivePlugin(async () => {
    const result = await relay.toolCallExecute('safe_tool', { value: 1 }, (args) => ({
      result: args,
      annotation: { source: 'application' },
    }));

    assert.deepEqual(result, {
      result: { value: 1, plugin_tag: 'documentation' },
      annotation: { source: 'application' },
    });
  });
});

test('tool policy blocks the configured tool', () => {
  const policy = registeredCallbacks().get('registerToolConditionalExecutionGuardrail');

  assert.equal(policy('dangerous_tool', { value: 1 }), "tool 'dangerous_tool' is blocked");
});

test('LLM requests are rewritten', async () => {
  await withActivePlugin(async () => {
    const result = await relay.llmCallExecute(
      'allowed-model',
      { headers: {}, content: { model: 'allowed-model' } },
      (request) => ({ headers: request.headers }),
    );

    assert.equal(result.headers['x-nemo-relay-plugin'], 'documentation');
  });
});

test('LLM policy blocks the configured model', () => {
  const policy = registeredCallbacks().get('registerLlmConditionalExecutionGuardrail');

  assert.equal(policy({ headers: {}, content: { model: 'restricted-model' } }), "model 'restricted-model' is blocked");
});

test('LLM stream chunks are transformed', async () => {
  const intercept = registeredCallbacks().get('registerLlmStreamExecutionIntercept');

  const transformed = intercept({ headers: {}, content: { model: 'allowed-model' } }, async () =>
    (async function* () {
      yield { chunk: 1 };
      yield { chunk: 2 };
    })(),
  );
  const chunks = [];
  for await (const chunk of transformed) chunks.push(chunk);

  assert.deepEqual(chunks, [
    { chunk: 1, plugin_stream: true },
    { chunk: 2, plugin_stream: true },
  ]);
});

test('subscriber observes an emitted event', async () => {
  await withActivePlugin(async () => {
    relay.event('documentation-event', null, { emitted: true });
    await relay.flushSubscribers();

    assert.ok(documentationPlugin.events.includes('documentation-event'));
  });
});

test('runtime controls emit a mark and an isolated scope only when enabled', async () => {
  await withActivePlugin(async () => {
    await relay.toolCallExecute('safe_tool', { value: 1 }, (args) => ({ result: args }));
    await relay.flushSubscribers();

    assert.ok(documentationPlugin.events.includes('documentation-plugin.request'));
    assert.ok(documentationPlugin.events.includes('documentation-plugin.isolated'));
  });

  const runtimeDisabled = config('enforce');
  runtimeDisabled.components[0].config.runtime = { emit_marks: false, emit_isolated_scope: false };
  await withActivePlugin(async () => {
    await relay.toolCallExecute('safe_tool', { value: 1 }, (args) => ({ result: args }));
    await relay.flushSubscribers();

    assert.ok(!documentationPlugin.events.includes('documentation-plugin.request'));
    assert.ok(!documentationPlugin.events.includes('documentation-plugin.isolated'));
  }, runtimeDisabled);
});

test('runtime controls do not depend on request rewriting', async () => {
  const requestsDisabled = config('enforce');
  requestsDisabled.components[0].config.requests.enabled = false;
  await withActivePlugin(async () => {
    await relay.toolCallExecute('safe_tool', { value: 1 }, (args) => ({ result: args }));
    await relay.flushSubscribers();

    assert.ok(documentationPlugin.events.includes('documentation-plugin.request'));
  }, requestsDisabled);
});

test('teardown removes the plugin kind', async () => {
  const restoreEnvironment = isolateExampleEnvironment();
  plugin.register('documentation-plugin', documentationPlugin);
  try {
    await plugin.initialize(config('enforce'));
    plugin.clear();
    assert.equal(plugin.deregister('documentation-plugin'), true);
    assert.equal(plugin.listKinds().includes('documentation-plugin'), false);
  } finally {
    plugin.clear();
    plugin.deregister('documentation-plugin');
    restoreEnvironment();
  }
});

test('registration rejects a duplicate kind and missing deregistration is false', () => {
  const restoreEnvironment = isolateExampleEnvironment();
  plugin.register('documentation-plugin', documentationPlugin);
  try {
    assert.throws(() => plugin.register('documentation-plugin', documentationPlugin));
    assert.equal(plugin.deregister('missing-documentation-plugin'), false);
  } finally {
    plugin.clear();
    plugin.deregister('documentation-plugin');
    restoreEnvironment();
  }
});
